platform/cxxsupport/mstd_span

```diff
@@ -0,0 +1,381 @@
+#include <mstd_type_traits>
+#include "mbed_assert.h"
+
+namespace mstd {
+
+constexpr size_t dynamic_extent = -1;
+
+template <typename, size_t = dynamic_extent> class span;
+
+namespace detail {
+
+template<typename ElementType, size_t Extent>
+class storage {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t) noexcept :
+        _data(ptr)
+    {}
+
+    ElementType *_data = nullptr;
+    static constexpr size_t _size = Extent;
+};
+
+template<typename ElementType>
+class storage<ElementType, dynamic_extent> {
+public:
+    constexpr storage() noexcept = default;
+
+    constexpr storage(ElementType *ptr, size_t size) noexcept :
+        _data(ptr), _size(size)
+    {}
+
+    ElementType *_data = nullptr;
+    size_t _size = 0;
+};
+
+template<typename>
+struct is_span: std::false_type {};
+
+template<typename T, size_t E>
+struct is_span<span<T, E>>: std::true_type {};
+
+template<typename>
+struct is_std_array: std::false_type {};
+
+template<typename T, size_t N>
+struct is_std_array<std::array<T, N>>: std::true_type {};
+
+template<typename, typename = void>
+struct has_size : std::false_type {};
+
+template<typename T>
+struct has_size<T, void_t<decltype(mstd::size(std::declval<T>()))>>:
+    std::true_type {};
+
+template<typename, typename = void>
+struct has_data : std::false_type {};
+
+template<typename T>
+struct has_data<T, void_t<decltype(mstd::data(std::declval<T>()))>>:
+    std::true_type {};
+
+template<typename T, typename U = mstd::remove_cvref_t<T>>
+struct is_container{
+    static constexpr bool value =
+        not is_span<U>::value && not is_std_array<U>::value &&
+            not std::is_array<U>::value && has_size<T>::value &&
+                has_data<T>::value;
+};
+
+template <class T>
+using iterator_t = decltype(mstd::begin(std::declval<T&>()));
+
+template <class R>
+using range_reference_t = mstd::iter_reference_t<iterator_t<R>>;
+
+template <typename, typename, typename = void>
+struct is_compatible : std::false_type {};
+
+template <typename T, typename E>
+struct is_compatible<T, E,
+    typename mstd::enable_if_t<not mstd::is_same<
+            typename mstd::remove_cv_t<
+                decltype(mstd::data(std::declval<T>()))
+            >, void>::value>>:
+mstd::is_convertible<remove_pointer_t<
+    decltype(mstd::data(std::declval<T>()))> (*)[], E (*)[]>{};
+
+} // namespace detail
+
+template<typename ElementType, size_t Extent>
+class span {
+public:
+    using element_type = ElementType;
+    using value_type = typename mstd::remove_cv_t<element_type>;
+    using index_type = size_t;
+    using difference_type = ptrdiff_t;
+    using pointer = element_type *;
+    using const_pointer = const element_type *;
+    using reference = element_type &;
+    using const_reference = const element_type &;
+    using iterator = pointer;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+
+    static constexpr index_type extent = Extent;
```

Comment:
Out-of-class definition also required for this. (Failures can occur at -O0, typically).